compiler/hash-ast/src/ast.rs

```diff
@@ -401,6 +402,111 @@ define_tree! {
         }
     }
 
+    /// Macro invocation argument.
+    #[derive(Debug, PartialEq, Clone)]
+    #[node]
+    pub struct MacroInvocationArg {
+        /// Optional name for the function argument, e.g `f(x = 3);`.
+        pub name: OptionalChild!(Name),
+
+        /// Each argument of the function call, as an expression.
+        pub value: Child!(Expr),
+    }
+
+    #[derive(Debug, PartialEq, Clone)]
+    #[node]
+    pub struct MacroInvocationArgs {
+        pub args: Children!(MacroInvocationArg),
+    }
+
+
+    #[derive(Debug, PartialEq, Clone)]
+    #[node]
+    pub struct MacroInvocation {
+        /// The name of the macro. @@Todo: make this into an access-name.
+        pub name: Child!(Name),
+
+        /// Any arguments to the macro itself.
+        pub args: OptionalChild!(MacroInvocationArgs),
+    }
+
+    /// This is a collection of macro invocations that can occur on a single
+    /// item which are collected into a single node.
+    ///
+    /// @@Note: the `AstNodeId` of the `MacroInvocations` will be the same as
+    /// the id of the children `invocations` node. The [`MacroInvocations`]
+    /// struct is only to effectively collect logic for handling clumps of such
+    /// invocations under one node.
+    #[derive(Debug, PartialEq, Clone)]
+    #[node]
+    pub struct MacroInvocations {
+        pub invocations: Children!(MacroInvocation),
+    }
+
+    impl MacroInvocations {
+        /// Create a new empty [MacroInvocations].
+        pub fn empty(span: Span) -> AstNode<Self> {
+            let contents = Self { invocations: AstNodes::empty(span) };
+            let id = contents.invocations.id;
+            AstNode::with_id(contents, id)
+        }
+
+        /// Get the number of invocations that are contained within this node.
+        pub fn len(&self) -> usize {
+            self.invocations.len()
+        }
+
+        /// Check whether this node contains any invocations.
+        pub fn is_empty(&self) -> bool {
+            self.invocations.is_empty()
+        }
+    }
+
+    /// A macro invocation on an expression.
+    #[derive(PartialEq, Debug, Clone)]
+    #[node]
+    pub struct ExprMacroInvocation {
+        /// The directives that apply on the subject expression.
+        pub macro_args: Child!(MacroInvocations),
+
+        /// The subjection expression of the invocation.
```

Comment:
Shouldn't this be either an expression or tokens, depending on the kind?

Reply:
Yeah, I'd like to make this change after the original change, consider some additional constrains about ast/token macros, and then write them down.